mayastor/src/host/device.rs

```diff
@@ -0,0 +1,124 @@
+use std::{
+    collections::HashMap,
+    ffi::{OsStr, OsString},
+    io::Error,
+};
+
+use proc_mounts::{MountInfo, MountIter};
+use rpc::mayastor::BlockDevice;
+use udev::{Device, Enumerator};
+
+fn unwrap(value: Option<&OsStr>) -> String {
+    String::from(value.map(|s| s.to_str()).flatten().unwrap_or(""))
+}
+
+fn map(value: Option<&OsStr>) -> Option<String> {
+    value.map(|s| s.to_str()).flatten().map(String::from)
+}
+
+fn parse(value: Option<&OsStr>) -> Option<u64> {
+    map(value).map(|s| s.parse().ok()).flatten()
+}
+
+fn new_device(
+    include: bool,
+    device: &Device,
+    mounts: &HashMap<OsString, MountInfo>,
+) -> Option<BlockDevice> {
+    if let Some(devname) = device.property_value("DEVNAME") {
+        let mut fstype = map(device.property_value("ID_FS_TYPE"));
+        let size: u64 = parse(device.attribute_value("size")).unwrap_or(0);
+        let mountinfo = mounts.get(devname);
+
+        if fstype.is_none() {
+            fstype = mountinfo.map(|m| m.fstype.clone());
+        }
+
+        let available =
+            include && size > 0 && fstype.is_none() && mountinfo.is_none();
+
+        return Some(BlockDevice {
+            devname: String::from(devname.to_str().unwrap_or("")),
+            devpath: unwrap(device.property_value("DEVPATH")),
+            devtype: unwrap(device.property_value("DEVTYPE")),
+            devmajor: parse(device.property_value("MAJOR")).unwrap_or(0),
+            devminor: parse(device.property_value("MINOR")).unwrap_or(0),
+            size,
+            fstype: fstype.unwrap_or_else(|| String::from("")),
+            mountpoint: mountinfo
+                .map(|m| String::from(m.dest.to_string_lossy()))
+                .unwrap_or_else(|| String::from("")),
+            available,
+        });
+    }
+    None
+}
+
+fn get_mounts() -> Result<HashMap<OsString, MountInfo>, Error> {
+    let mut table: HashMap<OsString, MountInfo> = HashMap::new();
+
+    for entry in MountIter::new()? {
+        if let Ok(mount) = entry {
+            table.insert(OsString::from(mount.source.clone()), mount);
+        }
+    }
+
+    Ok(table)
+}
+
+fn get_disks(
+    all: bool,
+    mounts: &HashMap<OsString, MountInfo>,
+) -> Result<Vec<BlockDevice>, Error> {
+    let mut list: Vec<BlockDevice> = Vec::new();
+
+    let mut enumerator = Enumerator::new()?;
+
+    enumerator.match_subsystem("block")?;
+    enumerator.match_property("DEVTYPE", "disk")?;
+
+    for entry in enumerator.scan_devices()? {
+        let partitions = get_partitions(true, &entry, mounts)?;
+        let include = partitions.is_empty();
+
+        if let Some(device) = new_device(include, &entry, &mounts) {
+            if all || device.available {
```

Comment:
could we just push devices to the list and just "filter" them on the return?

also rather than having the "manual check" I wonder if we can have a list of filters which is constructed after parsing the cli arguments.

Reply:
That is a nice idea, although given the current time constraints is probably something to look at further down the line.

Reply:
:+1: